File: DSAndML/Probability/probabilitymassfunc.py
def probabilitymassfunc(prob_list, tol=1e-9):
    total = sum(prob_list)
    print(f"Sum of probabilities: {total}")

    # Check if the sum of probabilities is close enough to 1, within a tolerance
    return abs(total - 1) < tol


prob_list = [0.5, 0.5]
print(probabilitymassfunc(prob_list))

"""
Tolerance Check (tol=1e-9): This allows for small floating-point differences, so instead of checking if sum(prob_list) == 1, it checks if the sum is close enough to 1 (within the tolerance of 1e-9).
abs(total - 1) < tol: This checks whether the difference between the sum and 1 is smaller than the tolerance, allowing for floating-point inaccuracies.
"""

# using numpy

import numpy as np


def probability_mass_func_numpy(prob_list):
    total = np.sum(prob_list)
    print(f"Sum of probabilities: {total}")

    # Use numpy's built-in tolerance for floating-point comparisons
    return np.isclose(total, 1)


print(probability_mass_func_numpy(prob_list))


# Check if Probabilities Are Between 0 and 1
def probability_mass_func(prob_list):
    # Ensure all probabilities are between 0 and 1
    if all(0 <= p <= 1 for p in prob_list):
        total = sum(prob_list)
        print(f"Sum of probabilities: {total}")

        # Check if the sum is 1
        return abs(total - 1) < 1e-9  # Adjusted tolerance
    else:
        return False


# Example usage:
prob_list = [0.4, 0.6]
print(probability_mass_func(prob_list))

# Using fractions.Fraction for Exact Arithmetic

from fractions import Fraction


def probability_mass_func_fraction(prob_list):
    prob_list_fraction = [Fraction(p).limit_denominator() for p in prob_list]
    total = sum(prob_list_fraction)

    print(f"Sum of probabilities (as fraction): {total}")

    return total == 1


print(probability_mass_func_fraction(prob_list))

# Normalize manually

def probability_mass_func_normalized(prob_list):
    # Normalize the probabilities by dividing by the sum
    total = sum(prob_list)
    normalized = [p / total for p in prob_list]

    # Check if the sum of normalized probabilities equals 1
    return sum(normalized) == 1

print(probability_mass_func_normalized(prob_list))

File: DSAndML/Probability/probabilitymassfunc.R
probability_mass_func <- function(prob_list, tol = 1e-9) {
    total <- sum(prob_list)
    cat("Sum of probabilities: ", total, "\n")

    # Check if all probabilities are between 0 and 1 and sum to 1
    all_valid <- all(prob_list >= 0 & prob_list <= 1)
    close_to_one <- abs(total - 1) < tol

    return(all_valid && close_to_one)
}
prob_list <- c(0.2, 0.3, 0.5)
print(probability_mass_func(prob_list))


# Using R's fraction Library for Exact Arithmetic
library(MASS)

probability_mass_func_fraction <- function(prob_list) {
    # Convert the list to fractions
    total <- sum(fractions(prob_list))
    cat("Sum of probabilities (as fraction):", total, "\n")

    return(total == 1)
}

print(probability_mass_func_fraction(prob_list))


# Manual Normalization and Validation
probability_mass_func_normalized <- function(prob_list) {
    total <- sum(prob_list)
    normalized <- prob_list / total

    # Check if normalized probabilities sum to 1
    valid_sum <- isTRUE(all.equal(sum(normalized), 1))

    # Check if all probabilities are between 0 and 1
    all_valid <- all(normalized >= 0 & normalized <= 1)

    return(all_valid && valid_sum)
}

print(probability_mass_func_normalized(prob_list)) # Should return TRUE
